balancer/catalyst/catalyst_balancer.go

```diff
@@ -0,0 +1,171 @@
+package catalyst
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sort"
+
+	"github.com/livepeer/catalyst-api/cluster"
+)
+
+type CataBalancer struct {
+	Nodes []Node
+}
+
+func (c *CataBalancer) Start(ctx context.Context) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *CataBalancer) UpdateMembers(ctx context.Context, members []cluster.Member) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *CataBalancer) GetBestNode(ctx context.Context, redirectPrefixes []string, playbackID, lat, lon, fallbackPrefix string) (string, string, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *CataBalancer) MistUtilLoadBalance(ctx context.Context, stream, lat, lon string) (string, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *CataBalancer) MistUtilLoadSource(ctx context.Context, stream, lat, lon string) (string, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (c *CataBalancer) MistUtilLoadStreamStats(ctx context.Context, stream string) error {
+	//TODO implement me
+	panic("implement me")
+}
+
+func NewBalancer() *CataBalancer {
+	return &CataBalancer{}
+}
+
+func (c *CataBalancer) UpdateNodes(id string, nodeMetrics NodeMetrics) {
+	// TODO change c.Nodes to a map so that we don't have to worry about the size of the slice changing?
+	for i := range c.Nodes {
+		if c.Nodes[i].ID == id {
+			c.Nodes[i].NodeMetrics = nodeMetrics
+			break
+		}
+	}
+}
+
+func (c *CataBalancer) UpdateStreams(id string, streams map[string]Stream) {
+	for i := range c.Nodes {
+		if c.Nodes[i].ID == id {
+			c.Nodes[i].Streams = streams
+			break
+		}
+	}
+}
+
+// TODO: This is temporary until we have the real struct definition
+type Node struct {
+	ID      string
+	Streams map[string]Stream // Stream ID -> Stream
+	NodeMetrics
+	GeoLatitude  float64
+	GeoLongitude float64
+}
+
+type Stream struct {
+	ID string
+}
+
+type NodeMetrics struct {
+	CPUUsagePercentage       int64
+	RAMUsagePercentage       int64
+	BandwidthUsagePercentage int64
+}
+
+// All of the scores are in the range 0-2, where:
+// 2 = Good
+// 1 = Okay
+// 0 = Bad
+type ScoredNode struct {
+	Score       int64
+	GeoScore    int64
+	GeoDistance float64
+	Node
+}
+
+func (n Node) HasStream(streamID string) bool {
+	_, ok := n.Streams[streamID]
+	return ok
+}
+
+func (n Node) GetLoadScore() int {
+	if n.CPUUsagePercentage > 85 || n.BandwidthUsagePercentage > 85 || n.RAMUsagePercentage > 85 {
+		return 0
+	}
+	if n.CPUUsagePercentage > 50 || n.BandwidthUsagePercentage > 50 || n.RAMUsagePercentage > 50 {
+		return 1
+	}
+	return 2
+}
+
+func SelectNode(nodes []Node, streamID string, requestLatitude, requestLongitude float64) (Node, error) {
+	if len(nodes) == 0 {
+		return Node{}, fmt.Errorf("no nodes to select from")
+	}
+
+	topNodes := selectTopNodes(nodes, streamID, requestLatitude, requestLongitude, 3)
+	return topNodes[rand.Intn(len(topNodes))].Node, nil
```

Comment:
I'd add some logging for live debugging/analysing selection mechanism:
```go
...
topNodes := selectTopNodes(nodes, streamID, requestLatitude, requestLongitude, 3)
glog.V(7).Infof("Selecting node: all available: %v", nodes)
glog.V(5).Infof("Selecting node: topNodes: %v", topNodes)
selected : = topNodes[rand.Intn(len(topNodes))].Node
glog.V(3).Infof("Selecting node: selected %v", selected)
return selected, nil
```

Reply:
Yep good idea 👍 just want to check @iameli do you think it'll be ok to use logging rather than prometheus metrics? I'm just wondering if we're normally ok with those sorts of volumes of logging, it'll only be called once per playback session right so not too bad?